File: anvil-api/src/main/java/org/anvilpowered/anvil/api/Anvil.java
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.anvil.api;

import com.google.common.base.Preconditions;
import com.google.inject.Binder;
import com.google.inject.Binding;
import com.google.inject.Injector;
import com.google.inject.Module;
import org.anvilpowered.anvil.api.core.coremember.CoreMemberManager;
import org.anvilpowered.anvil.api.core.coremember.repository.CoreMemberRepository;
import org.anvilpowered.anvil.api.data.registry.Registry;
import org.anvilpowered.anvil.api.misc.BindingExtensions;
import org.anvilpowered.anvil.base.plugin.BasePlugin;

import java.util.HashMap;
import java.util.Map;

@SuppressWarnings({"unused"})
public class Anvil extends BasePlugin {

    static final Map<Long, Binding<?>> bindingsCache = new HashMap<>();
    protected static ServiceManager serviceManager;
    protected static Environment environment;
    private static final String NOT_LOADED = "Anvil has not been loaded yet!";

    Anvil(String name, Injector rootInjector, Module module) {
        super(name, rootInjector, module);
    }

    public static BindingExtensions getBindingExtensions(Binder binder) {
        return getServiceManager().provide(BindingExtensions.class, binder);
    }

    public static Environment.Builder getEnvironmentBuilder() {
        return getServiceManager().provide(Environment.Builder.class);
    }

    public static EnvironmentManager getEnvironmentManager() {
        return getServiceManager().provide(EnvironmentManager.class);
    }

    public static Environment getEnvironment() {
        return Preconditions.checkNotNull(environment, NOT_LOADED);
    }

    public static Platform getPlatform() {
        return getEnvironment().getInjector().getInstance(Platform.class);
    }

    public static Registry getRegistry() {
        return getEnvironment().getInjector().getInstance(Registry.class);
    }

    public static CoreMemberManager getCoreMemberManager() {
        return getEnvironment().getInjector().getInstance(CoreMemberManager.class);
    }

    public static CoreMemberRepository<?, ?> getCoreMemberRepository() {
        return getCoreMemberManager().getPrimaryComponent();
    }

    public static ServiceManager getServiceManager() {
        if (serviceManager != null) {
            return serviceManager;
        }
        try {
            return serviceManager = (ServiceManager)
                Class.forName("org.anvilpowered.anvil.api.ServiceManagerImpl").newInstance();
        } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
            throw new IllegalStateException("Could not find ServiceManager implementation!", e);
        }
    }
}

File: api/src/main/kotlin/org/anvilpowered/anvil/api/Anvil.kt
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020-2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.anvilpowered.anvil.api

import com.google.inject.Binding
import com.google.inject.Injector
import com.google.inject.Module
import org.anvilpowered.anvil.api.environment.Environment
import org.anvilpowered.anvil.api.environment.EnvironmentManager
import org.anvilpowered.anvil.base.plugin.BasePlugin
import org.anvilpowered.anvil.api.registry.Registry
import kotlin.experimental.ExperimentalTypeInference

open class Anvil constructor(name: String, rootInjector: Injector, module: Module) :
    BasePlugin(name, rootInjector, module) {

    companion object {
        val bindingsCache: MutableMap<Long, Binding<*>> = HashMap()

        var serviceManager: ServiceManager? = null
            get() {
                return field
                    ?: try {
                        (Class.forName(
                            "org.anvilpowered.anvil.core.ServiceManagerImpl"
                        ).getConstructor().newInstance() as ServiceManager).also {
                            serviceManager = it
                        }
                    } catch (e: InstantiationException) {
                        throw IllegalStateException("Could not find ServiceManager implementation!", e)
                    } catch (e: IllegalAccessException) {
                        throw IllegalStateException("Could not find ServiceManager implementation!", e)
                    } catch (e: ClassNotFoundException) {
                        throw IllegalStateException("Could not find ServiceManager implementation!", e)
                    }
            }

        var environment: Environment? = null
            get() = field ?: throw java.lang.IllegalStateException(NOT_LOADED)

        private const val NOT_LOADED = "Anvil has not been loaded yet!"

        @OptIn(ExperimentalTypeInference::class)
        inline fun environmentBuilder(@BuilderInference block: Environment.Builder.() -> Unit): Environment.Builder {
            return serviceManager!!.provide(Environment.Builder::class.java).apply(block)

        }

        val environmentManager: EnvironmentManager by lazy {
            serviceManager!!.provide(EnvironmentManager::class.java)
        }

        val platform: Platform by lazy {
            environment!!.injector.getInstance(Platform::class.java)
        }

        val registry: Registry by lazy {
            environment!!.injector.getInstance(Registry::class.java)
        }
    }
}
